# Laboratorio 2
Andrea Reyes -
20190265

In [30]:
import numpy as np
import pandas as pd
heroes = pd.read_csv("heroes_information.csv")

In [31]:
###instalar desde anaconda navigator######
from pandasql import *

In [32]:
pysqldf = lambda q: sqldf(q,globals())

In [33]:
query2 = " SELECT * FROM heroes limit 10"

In [34]:
pysqldf(query2)

id           name  Gender Eye color               Race Hair color  Height  \
0   0         A-Bomb    Male    yellow              Human    No Hair   203.0   
1   1     Abe Sapien    Male      blue      Icthyo Sapien    No Hair   191.0   
2   2       Abin Sur    Male      blue            Ungaran    No Hair   185.0   
3   3    Abomination    Male     green  Human / Radiation    No Hair   203.0   
4   4        Abraxas    Male      blue      Cosmic Entity      Black   -99.0   
5   5  Absorbing Man    Male      blue              Human    No Hair   193.0   
6   6    Adam Monroe    Male      blue               None      Blond   -99.0   
7   7   Adam Strange    Male      blue              Human      Blond   185.0   
8   8       Agent 13  Female      blue               None      Blond   173.0   
9   9      Agent Bob    Male     brown              Human      Brown   178.0   

           Publisher Skin color Alignment  Weight  
0      Marvel Comics       None      good   441.0  
1  Dark Horse Comics       blue      good    65.0  
2          DC Comics        red      good    90.0  
3      Marvel Comics       None       bad   441.0  
4      Marvel Comics       None       bad   -99.0  
5      Marvel Comics       None       bad   122.0  
6       NBC - Heroes       None      good   -99.0  
7          DC Comics       None      good    88.0  
8      Marvel Comics       None      good    61.0  
9      Marvel Comics       None      good    81.0



### 1. Cuál es el nombre del primer superhéroe de la tabla heroes?

In [35]:
p1 = "SELECT name FROM heroes"
pysqldf(p1)

name
0             A-Bomb
1         Abe Sapien
2           Abin Sur
3        Abomination
4            Abraxas
..               ...
729  Yellowjacket II
730             Ymir
731             Yoda
732          Zatanna
733             Zoom

[734 rows x 1 columns]

R// A-Bomb

### 2. Cuántas casas publicadoras existen?

In [36]:
p2 = "SELECT DISTINCT Publisher FROM heroes"
pysqldf(p2)

Publisher
0       Marvel Comics
1   Dark Horse Comics
2           DC Comics
3        NBC - Heroes
4           Wildstorm
5        Image Comics
6                None
7         Icon Comics
8                SyFy
9       Hanna-Barbera
10       George Lucas
11       Team Epic TV
12         South Park
13      HarperCollins
14        ABC Studios
15  Universal Studios
16          Star Trek
17     IDW Publishing
18           Shueisha
19      Sony Pictures
20      J. K. Rowling
21        Titan Books
22          Rebellion
23          Microsoft
24   J. R. R. Tolkien

R// 25 casas publicadoras

### 3. Cuántos superhéroes miden más de 2 metros (200 cms)

In [37]:
p3 = "SELECT COUNT(Id) FROM heroes WHERE Height > 200"
pysqldf(p3)

COUNT(Id)
0         59

R// 59

### 4. Cuántos superhéroes son humanos y miden más de 2 metros (200 cms)?

In [38]:
p4 = "SELECT * FROM heroes WHERE Race IN ('Human') AND Height > 200"
pysqldf(p4)

id           name  Gender Eye color   Race Hair color  Height  \
0     0         A-Bomb    Male    yellow  Human    No Hair   203.0   
1    29        Ant-Man    Male      blue  Human      Blond   211.0   
2    59           Bane    Male      None  Human       None   203.0   
3   119       Bloodaxe  Female      blue  Human      Brown   218.0   
4   221    Doctor Doom    Male     brown  Human      Brown   201.0   
5   373     Juggernaut    Male      blue  Human        Red   287.0   
6   391        Kingpin    Male      blue  Human    No Hair   201.0   
7   412         Lizard    Male       red  Human    No Hair   203.0   
8   476  Mr Incredible    Male      blue  Human      Blond   201.0   
9   557            Rey  Female     hazel  Human      Brown   297.0   
10  583       Scorpion    Male     brown  Human      Brown   211.0   
11  591       She-Hulk  Female     green  Human      Green   201.0   

            Publisher Skin color Alignment  Weight  
0       Marvel Comics       None      good   441.0  
1       Marvel Comics       None      good   122.0  
2           DC Comics       None       bad   180.0  
3       Marvel Comics       None       bad   495.0  
4       Marvel Comics       None       bad   187.0  
5       Marvel Comics       None   neutral   855.0  
6       Marvel Comics       None       bad   203.0  
7       Marvel Comics       None       bad   230.0  
8   Dark Horse Comics       None      good   158.0  
9        George Lucas       None      good   -99.0  
10      Marvel Comics       None       bad   310.0  
11      Marvel Comics       None      good   315.0

R// 12

### 5. Cuántos superhéroes pesan más de 100 lbs y menos de 200 lbs?

In [39]:
p5 = "SELECT Name FROM heroes WHERE Weight BETWEEN 100 AND 200"
#p5 = "SELECT * FROM heroes WHERE Race IN(1,3,5,7)"
pysqldf(p5)

name
0    Absorbing Man
1       Agent Zero
2       Air-Walker
3            Alien
4            Amazo
..             ...
93         Warpath
94  Winter Soldier
95       Wolverine
96      Wonder Man
97  Wyatt Wingfoot

[98 rows x 1 columns]

R// 98

### 6. Cuántos superhéroes tienen los ojos de color azul o rojo?

In [40]:
p6 = "SELECT Name, `Eye color` FROM heroes WHERE `Eye color` IN ('red', 'blue')"
pysqldf(p6)

name Eye color
0         Abe Sapien      blue
1           Abin Sur      blue
2            Abraxas      blue
3      Absorbing Man      blue
4        Adam Monroe      blue
..               ...       ...
266      Yellow Claw      blue
267     Yellowjacket      blue
268  Yellowjacket II      blue
269          Zatanna      blue
270             Zoom       red

[271 rows x 2 columns]

R// 271

### 7. Cuántos superhéroes son Human, Mutant y tienen el pelo color Verde o son Vampiros con pelo Negro

In [41]:
p7 = "SELECT Name, Race, `Hair color` FROM heroes  WHERE Race IN ('Human','Mutant') AND `Hair color` IN ('Green') OR Race IN ('Vampire') AND `Hair color` IN ('Black')"
pysqldf(p7)

name     Race Hair color
0  Beast Boy    Human      Green
1      Blade  Vampire      Black
2      Joker    Human      Green
3    Polaris   Mutant      Green
4   She-Hulk    Human      Green

R// 5

### 8. Cuál es el nombre del primer superhéroe si ordeno la tabla por raza en orden descendente?

In [42]:
p8 = "SELECT * FROM heroes ORDER BY Race DESC"
#p8 = "SELECT * FROM heroes GROUP BY "
pysqldf(p8)

id            name  Gender Eye color             Race Hair color  \
0    611  Solomon Grundy    Male     black           Zombie      White   
1    274          Gamora  Female    yellow    Zen-Whoberian      Black   
2    731            Yoda    Male     brown   Yoda's species      White   
3    527        Predator    Male      None           Yautja       None   
4     17           Alien    Male      None  Xenomorph XX121    No Hair   
..   ...             ...     ...       ...              ...        ...   
729  723          Wondra  Female      None             None       None   
730  724  Wyatt Wingfoot    Male     brown             None      Black   
731  726           X-Man    Male      blue             None      Brown   
732  727     Yellow Claw    Male      blue             None    No Hair   
733  733            Zoom    Male       red             None      Brown   

     Height          Publisher Skin color Alignment  Weight  
0     279.0          DC Comics       None       bad   437.0  
1     183.0      Marvel Comics      green      good    77.0  
2      66.0       George Lucas      green      good    17.0  
3     213.0  Dark Horse Comics       None       bad   234.0  
4     244.0  Dark Horse Comics      black       bad   169.0  
..      ...                ...        ...       ...     ...  
729   -99.0      Marvel Comics       None      good   -99.0  
730   196.0      Marvel Comics       None      good   117.0  
731   175.0      Marvel Comics       None      good    61.0  
732   188.0      Marvel Comics       None       bad    95.0  
733   185.0          DC Comics       None       bad    81.0  

[734 rows x 11 columns]

R// Solomon Grundy

### 9. Cuántos superhéroes son de género masculino y cuántas de género femenino?

In [43]:
p9 = "SELECT Count(*), Gender FROM heroes Group by Gender"
pysqldf(p9)

Count(*)  Gender
0        29    None
1       200  Female
2       505    Male

R// Femenino = 200 - Masculino = 505

### 10. Cuántas casas publicadoras tienen más de 15 superhéroes?

In [44]:
p10 = "SELECT Count(*), Publisher FROM heroes Group by Publisher HAVING COUNT(*) > 15"
pysqldf(p10)

Count(*)          Publisher
0       215          DC Comics
1        18  Dark Horse Comics
2       388      Marvel Comics
3        19       NBC - Heroes

R// 4